File: ios/GuideStreamTV/Views/GenreTileGrid.swift
//
//  GenreTileGrid.swift
//  GuideStreamTV
//
//  The genre entry point for Search & Browse: artwork tiles, two up.
//
//  A tinted square with an SF Symbol (the Home treatment) does not sell a
//  category — the tile has to show what is inside it. Each tile pulls the top
//  backdrop for its genre once per session and falls back to its brand tint
//  until that lands, so the grid is never empty and never janks.
//

import SwiftUI

// MARK: - Tint

/// Per-genre gradient, used as the tile's resting state and as the fallback
/// behind artwork that has not loaded yet.
enum BrowseTint {
    private static func rgb(_ hex: UInt32) -> Color {
        Color(
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255
        )
    }

    private static let ramps: [String: (UInt32, UInt32)] = [
        "crime":         (0xC0392B, 0x2B0A0E),
        "scifi":         (0x1A6FE8, 0x06111F),
        "horror":        (0x5A1416, 0x050203),
        "anime":         (0x6C3BF5, 0x00C2E0),
        "comedy":        (0x2ECC71, 0x07160E),
        "drama":         (0x8E44AD, 0x120620),
        "action":        (0xF5821F, 0x1C0E02),
        "documentary":   (0x00A99D, 0x02171A),
        "romance":       (0xFF4785, 0x1C0413),
        "international": (0x3D7EA6, 0x050D18)
    ]

    static func colors(for genreId: String) -> [Color] {
        let ramp = ramps[genreId] ?? (0x2D1454, 0x04090F)
        return [rgb(ramp.0), rgb(ramp.1)]
    }
}

// MARK: - Artwork

/// Session cache of one backdrop per genre.
///
/// Ten concurrent discover calls, once, the first time a browse surface
/// appears. Results are held for the life of the process — the tiles are
/// decoration, so a stale-by-an-hour backdrop is not worth a refetch.
@MainActor
@Observable
final class BrowseArtworkStore {
    static let shared = BrowseArtworkStore()

    private(set) var backdrops: [String: String] = [:]
    private var isLoading = false

    private init() {}

    func loadIfNeeded() async {
        guard !isLoading, backdrops.count < BrowseCatalog.genres.count else { return }
        isLoading = true
        defer { isLoading = false }

        let pending = BrowseCatalog.genres.filter { backdrops[$0.id] == nil }
        let fetched = await withTaskGroup(of: (String, String?).self) { group in
            for genre in pending {
                group.addTask {
                    // No provider filter here: the tile should show the genre's
                    // best-known title, not whatever the user happens to have.
                    let filters = BrowseFilters(genreIds: [genre.id], onlyMyServices: false)
                    let page = try? await TMDBService.shared.discoverBrowse(filters)
                    let hero = page?.results.first { $0.backdropPath != nil }
                    return (genre.id, hero?.backdropUrl)
                }
            }
            var out: [String: String] = [:]
            for await (id, url) in group {
                if let url { out[id] = url }
            }
            return out
        }

        backdrops.merge(fetched) { _, new in new }
    }
}

// MARK: - Grid

struct GenreTileGrid: View {
    var onSelect: (BrowseGenre) -> Void

    @State private var artwork = BrowseArtworkStore.shared

    private let columns = [
        GridItem(.flexible(), spacing: 10),
        GridItem(.flexible(), spacing: 10)
    ]

    var body: some View {
        LazyVGrid(columns: columns, spacing: 10) {
            ForEach(BrowseCatalog.genres) { genre in
                Button {
                    onSelect(genre)
                } label: {
                    tile(genre)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Browse \(genre.name)")
            }
        }
        .task { await artwork.loadIfNeeded() }
    }

    private func tile(_ genre: BrowseGenre) -> some View {
        let tint = BrowseTint.colors(for: genre.id)
        return ZStack(alignment: .bottomLeading) {
            RemoteImage(
                url: artwork.backdrops[genre.id].flatMap(URL.init(string:)),
                contentMode: .fill,
                fallbackColors: tint
            )
            .frame(maxWidth: .infinity)
            .frame(height: 84)
            .clipped()

            LinearGradient(
                colors: [Color.black.opacity(0), Color.black.opacity(0.72)],
                startPoint: .top,
                endPoint: .bottom
            )

            Text(genre.name)
                .scaledFont(size: 13, weight: .bold)
                .foregroundStyle(.white)
                .lineLimit(1)
                .minimumScaleFactor(0.85)
                .shadow(color: .black.opacity(0.6), radius: 6, y: 1)
                .padding(10)
        }
        .frame(height: 84)
        .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: 12, style: .continuous)
                .stroke(Color.white.opacity(0.08), lineWidth: 0.5)
        )
    }
}

// MARK: - Bridge

extension SearchResult {
    /// Adapts a raw discover result to the row/cell model the search surfaces
    /// already render. Browse has no per-title provider lookup — that would be
    /// one request per poster — so the service fields stay neutral.
    init(browseResult item: TMDBResult) {
        self.init(
            id: item.id,
            title: item.displayName,
            isTV: item.isTV,
            posterUrl: item.posterUrl,
            backdropUrl: item.backdropUrl,
            year: item.year,
            genreNames: [],
            serviceName: nil,
            serviceColor: Color(white: 0.18),
            serviceShort: ""
        )
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/browse/GenreTileGrid.kt
package com.rork.guidestreamtvandroid.ui.browse

import androidx.compose.foundation.background
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.clickable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rork.guidestreamtvandroid.data.models.BrowseCatalog
import com.rork.guidestreamtvandroid.data.models.BrowseFilters
import com.rork.guidestreamtvandroid.data.models.BrowseGenre
import com.rork.guidestreamtvandroid.data.remote.TMDBService
import com.rork.guidestreamtvandroid.ui.components.RemoteImage
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * The genre entry point for Search & Browse: artwork tiles, two up.
 * Mirrors iOS `GenreTileGrid.swift`.
 *
 * A tinted square with an icon (the Home treatment) does not sell a category —
 * the tile has to show what is inside it. Each tile pulls the top backdrop for
 * its genre once per process and falls back to its brand tint until that lands,
 * so the grid is never empty and never janks.
 */

// ---------------------------------------------------------------------------
// Tint
// ---------------------------------------------------------------------------

/**
 * Per-genre gradient, used as the tile's resting state and as the fallback
 * behind artwork that has not loaded yet.
 */
object BrowseTint {
    private val ramps: Map<String, Pair<Long, Long>> = mapOf(
        "crime" to (0xFFC0392B to 0xFF2B0A0E),
        "scifi" to (0xFF1A6FE8 to 0xFF06111F),
        "horror" to (0xFF5A1416 to 0xFF050203),
        "anime" to (0xFF6C3BF5 to 0xFF00C2E0),
        "comedy" to (0xFF2ECC71 to 0xFF07160E),
        "drama" to (0xFF8E44AD to 0xFF120620),
        "action" to (0xFFF5821F to 0xFF1C0E02),
        "documentary" to (0xFF00A99D to 0xFF02171A),
        "romance" to (0xFFFF4785 to 0xFF1C0413),
        "international" to (0xFF3D7EA6 to 0xFF050D18),
    )

    fun brush(genreId: String): Brush {
        val ramp = ramps[genreId] ?: (0xFF2D1454 to 0xFF04090F)
        return Brush.linearGradient(listOf(Color(ramp.first), Color(ramp.second)))
    }
}

// ---------------------------------------------------------------------------
// Artwork
// ---------------------------------------------------------------------------

/**
 * Process-wide cache of one backdrop per genre.
 *
 * Ten concurrent discover calls, once, the first time a browse surface appears.
 * The tiles are decoration, so a stale-by-an-hour backdrop is not worth a
 * refetch.
 */
object BrowseArtworkStore {
    private val _backdrops = MutableStateFlow<Map<String, String>>(emptyMap())
    val backdrops: StateFlow<Map<String, String>> = _backdrops.asStateFlow()

    private val lock = Mutex()

    /**
     * How many backdrops each genre offers up for the assignment pass. One was
     * not enough: a title that tops two genres claimed both tiles.
     */
    private const val CANDIDATE_DEPTH = 8

    suspend fun loadIfNeeded() {
        if (_backdrops.value.size >= BrowseCatalog.genres.size) return
        lock.withLock {
            val current = _backdrops.value
            val pending = BrowseCatalog.genres.filter { current[it.id] == null }
            if (pending.isEmpty()) return
            val tmdb = TMDBService.get()
            val candidates: Map<String, List<String>> = coroutineScope {
                pending.map { genre ->
                    async {
                        // No provider filter here: the tile should show the
                        // genre's best-known title, not whatever the user has.
                        val page = tmdb.discoverBrowse(
                            BrowseFilters(genreIds = setOf(genre.id), onlyMyServices = false)
                        )
                        genre.id to page.results
                            .mapNotNull { it.backdropUrl }
                            .take(CANDIDATE_DEPTH)
                    }
                }.map { it.await() }.toMap()
            }

            // Assign in catalogue order rather than in completion order, so
            // which tile gets first claim on a shared title does not depend on
            // which network call happened to return first.
            //
            // Each tile takes its most popular backdrop that no earlier tile
            // has taken. Reacher tops both Crime & Thriller and Action, and the
            // two tiles used to show the identical image.
            val used = current.values.toMutableSet()
            val resolved = mutableMapOf<String, String>()
            for (genre in BrowseCatalog.genres) {
                val options = candidates[genre.id].orEmpty()
                if (options.isEmpty()) continue
                // If every candidate is already spoken for, take the first
                // anyway: a repeated tile still reads better than an empty one.
                val pick = options.firstOrNull { it !in used } ?: options[0]
                used += pick
                resolved[genre.id] = pick
            }

            _backdrops.value = current + resolved
        }
    }
}

// ---------------------------------------------------------------------------
// Grid
// ---------------------------------------------------------------------------

@Composable
fun GenreTileGrid(
    onSelect: (BrowseGenre) -> Unit,
    modifier: Modifier = Modifier,
) {
    val backdrops by BrowseArtworkStore.backdrops.collectAsStateWithLifecycle()
    LaunchedEffect(Unit) { BrowseArtworkStore.loadIfNeeded() }

    // A plain Column of Rows, not a LazyVerticalGrid: this sits inside another
    // scrolling container, and nesting lazy grids crashes at measure time.
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        BrowseCatalog.genres.chunked(2).forEach { row ->
            Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                row.forEach { genre ->
                    GenreTile(
                        genre = genre,
                        backdropUrl = backdrops[genre.id],
                        onClick = { onSelect(genre) },
                        modifier = Modifier.weight(1f),
                    )
                }
                // Keeps a lone trailing tile at half width instead of stretching.
                if (row.size == 1) Box(modifier = Modifier.weight(1f))
            }
        }
    }
}

@Composable
private fun GenreTile(
    genre: BrowseGenre,
    backdropUrl: String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .height(84.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(BrowseTint.brush(genre.id))
            .clickable(
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
            ) { onClick() },
        contentAlignment = Alignment.BottomStart,
    ) {
        if (backdropUrl != null) {
            RemoteImage(
                url = backdropUrl,
                contentDescription = genre.displayName,
                modifier = Modifier.fillMaxSize(),
                cornerRadius = 12,
            )
        }

        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(
                    Brush.verticalGradient(
                        listOf(Color.Transparent, Color.Black.copy(alpha = 0.72f))
                    )
                )
        )

        Text(
            text = genre.displayName,
            fontSize = 13.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.padding(10.dp),
        )
    }
}
